include/opt-sched/Scheduler/aco.h

```diff
@@ -29,22 +31,23 @@ class ACOScheduler : public ConstrainedScheduler {
   virtual ~ACOScheduler();
   FUNC_RESULT FindSchedule(InstSchedule *schedule, SchedRegion *region);
   inline void UpdtRdyLst_(InstCount cycleNum, int slotNum);
-  // Set the initial schedule for ACO 
+  // Set the initial schedule for ACO
   // Default is NULL if none are set.
   void setInitialSched(InstSchedule *Sched);
-  
+
 private:
   pheremone_t &Pheremone(SchedInstruction *from, SchedInstruction *to);
   pheremone_t &Pheremone(InstCount from, InstCount to);
   double Score(SchedInstruction *from, Choice choice);
 
   void PrintPheremone();
 
-  SchedInstruction *SelectInstruction(std::vector<Choice> ready,
-                                      SchedInstruction *lastInst);
+  SchedInstruction *
+  SelectInstruction(const llvm::SmallVectorImpl<Choice> &ready,
+                    SchedInstruction *lastInst);
   void UpdatePheremone(InstSchedule *schedule);
-  InstSchedule *FindOneSchedule();
-  pheremone_t *pheremone_;
+  std::unique_ptr<InstSchedule> FindOneSchedule();
+  llvm::SmallVector<pheremone_t, 0> pheremone_;
```

Comment:
Is there any reasonable guess about the average size of pheremone or does it really make sense to have no space reserved?

Reply:
pheremone_ has a size of (instruction count)*(instruction count+1), It is usually a largeish allocation.  It is highly variable in size depending on the scheduling region.

Reply:
If we have an idea of the distribution in scheduling region size, it could make sense to provide the small size as @kerbowa says. For example, if a block size of 4 instructions was relatively common, providing a small size of `4 * 5 == 20` would remove the allocation for regions of size 4 or smaller.

In this case, I don't think we need to worry about this. The allocation is unlikely to be a bottleneck in our code, and we can't know if it is unless we have profiled our code for small scheduling regions. Since we have not done so, I think we should leave it as `0`; potential future optimization work could use a small size of `0` as a sign that there may be some optimizing opportunity.

Reply:
I agree it's unlikely to be a bottleneck, so lets leave it as is.
For awareness, in LLVM we normally pre-allocate some amount of data in the SmallVector object even if you are just guessing. The convention is to use powers of 2. `4` `8` `16` `32` ect..